File: frontend/src/api/autopilot.ts
/**
 * ADs Growth System - Autopilot API Hooks
 *
 * React Query hooks for autopilot action management:
 * - Action queue
 * - Approval workflow
 * - Status monitoring
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiClient } from './client';

// =============================================================================
// Types
// =============================================================================

export type ActionStatus =
  | 'queued'
  | 'pending_approval'
  | 'approved'
  | 'applied'
  | 'failed'
  | 'dismissed';

export type ActionType =
  | 'budget_increase'
  | 'budget_decrease'
  | 'pause_campaign'
  | 'pause_adset'
  | 'pause_creative'
  | 'enable_campaign'
  | 'enable_adset'
  | 'enable_creative'
  | 'bid_increase'
  | 'bid_decrease';

export interface ActionApprover {
  id: number;
  name: string | null;
  department: string | null;
}

export interface AutopilotAction {
  id: string;
  date: string;
  action_type: ActionType;
  entity_type: 'campaign' | 'adset' | 'creative';
  entity_id: string;
  entity_name: string | null;
  platform: string;
  action_json: Record<string, unknown>;
  before_value: Record<string, unknown> | null;
  after_value: Record<string, unknown> | null;
  status: ActionStatus;
  created_at: string;
  approved_at: string | null;
  applied_at: string | null;
  error: string | null;
  /** True when the enforcement gate soft-blocked this action and an
   *  operator confirmation is required to execute it. */
  requires_confirmation: boolean;
  confirmation_token: string | null;
  /** Who accepted this action (accountability surface); null until approved. */
  approved_by: ActionApprover | null;
}

export interface AutopilotStatus {
  autopilot_level: number;
  autopilot_level_name: string;
  pending_actions: number;
  caps: {
    max_daily_budget_change: number;
    max_budget_pct_change: number;
    max_actions_per_day: number;
  };
  enabled: boolean;
}

export interface ActionsSummary {
  days: number;
  start_date: string;
  status_counts: Record<ActionStatus, number>;
  type_counts: Record<ActionType, number>;
  platform_counts: Record<string, number>;
  total: number;
  pending_approval: number;
  success_rate: number;
}

export interface QueueActionRequest {
  action_type: ActionType;
  entity_type: 'campaign' | 'adset' | 'creative';
  entity_id: string;
  entity_name: string;
  platform: string;
  action_json: Record<string, unknown>;
  before_value?: Record<string, unknown>;
}

// =============================================================================
// Query Hooks
// =============================================================================

/**
 * Get current autopilot status for a tenant.
 */
export function useAutopilotStatus(tenantId: number) {
  return useQuery({
    queryKey: ['autopilot-status', tenantId],
    queryFn: async () => {
      const response = await apiClient.get<{ data: AutopilotStatus }>(`/autopilot/status`);
      return response.data.data;
    },
    enabled: !!tenantId,
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 30 * 1000, // Refetch every 30 seconds
  });
}

/**
 * Get actions for a tenant.
 */
export function useAutopilotActions(
  tenantId: number,
  filters?: {
    date?: string;
    status?: ActionStatus;
    platform?: string;
    limit?: number;
  }
) {
  return useQuery({
    queryKey: ['autopilot-actions', tenantId, filters],
    queryFn: async () => {
      const params = new URLSearchParams();
      if (filters?.date) params.append('date', filters.date);
      if (filters?.status) params.append('status', filters.status);
      if (filters?.platform) params.append('platform', filters.platform);
      if (filters?.limit) params.append('limit', filters.limit.toString());

      const response = await apiClient.get<{
        data: { actions: AutopilotAction[]; count: number };
      }>(`/autopilot/actions?${params}`);
      return response.data.data;
    },
    enabled: !!tenantId,
    staleTime: 30 * 1000, // 30 seconds
    refetchInterval: 60 * 1000, // Refetch every minute
  });
}

/**
 * Get actions summary for a tenant.
 */
export function useActionsSummary(tenantId: number, days: number = 7) {
  return useQuery({
    queryKey: ['autopilot-summary', tenantId, days],
    queryFn: async () => {
      const response = await apiClient.get<{ data: ActionsSummary }>(
        `/autopilot/actions/summary?days=${days}`
      );
      return response.data.data;
    },
    enabled: !!tenantId,
    staleTime: 5 * 60 * 1000, // 5 minutes
  });
}

// =============================================================================
// Outcome Summary — powers the upgrade nudges
// =============================================================================

export type OutcomePeriod = '24h' | '7d' | '30d';

export interface OutcomeSummary {
  period_start: string;
  period_end: string;
  total_value_cents: number;
  decisions_count: number;
  saved_cents: number;
  earned_cents: number;
  prevented_cents: number;
  confidence_breakdown: Record<'low' | 'medium' | 'high', number>;
}

/**
 * Aggregate outcomes for the current tenant over a period. Drives the
 * OutcomeNudge upgrade trigger — only renders when total_value_cents
 * crosses the eligibility threshold + tenant is on Starter/trial.
 *
 * Phase A returns 0/0/0 for everyone (stub estimator); Phase B swaps
 * in counterfactual numbers.
 */
export function useAutopilotOutcomeSummary(tenantId: number, period: OutcomePeriod = '7d') {
  return useQuery({
    queryKey: ['autopilot-outcomes', tenantId, period],
    queryFn: async () => {
      const response = await apiClient.get<{ data: OutcomeSummary }>(
        `/autopilot/outcomes/summary?period=${period}`
      );
      return response.data.data;
    },
    enabled: !!tenantId,
    // Cheap to recompute on the backend (single aggregate query); refresh
    // every 5 min so a fresh autopilot decision can update the nudge
    // without a full page reload.
    staleTime: 5 * 60 * 1000,
    refetchInterval: 5 * 60 * 1000,
  });
}

/**
 * Get a specific action by ID.
 */
export function useAutopilotAction(tenantId: number, actionId: string) {
  return useQuery({
    queryKey: ['autopilot-action', tenantId, actionId],
    queryFn: async () => {
      const response = await apiClient.get<{ data: { action: AutopilotAction } }>(
        `/autopilot/actions/${actionId}`
      );
      return response.data.data.action;
    },
    enabled: !!tenantId && !!actionId,
  });
}

// =============================================================================
// Mutation Hooks
// =============================================================================

/**
 * Queue a new action.
 */
export function useQueueAction(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (request: QueueActionRequest) => {
      const response = await apiClient.post<{
        data: {
          action: AutopilotAction;
          auto_approved: boolean;
          requires_approval: boolean;
          reason: string | null;
        };
      }>(`/autopilot/actions`, request);
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-summary', tenantId] });
    },
  });
}

/**
 * Approve an action.
 */
export function useApproveAction(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (actionId: string) => {
      const response = await apiClient.post<{ data: { action: AutopilotAction } }>(
        `/autopilot/actions/${actionId}/approve`
      );
      return response.data.data.action;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
    },
  });
}

/**
 * Confirm a soft-blocked action (enforcement override) and execute it.
 *
 * The backend consumes the action's stored confirmation token, audit-logs
 * the override, and dispatches execution. Hard-blocked actions cannot be
 * confirmed.
 */
export function useConfirmAction(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async ({
      actionId,
      overrideReason,
    }: {
      actionId: string;
      overrideReason?: string;
    }) => {
      const response = await apiClient.post<{
        data: { action: AutopilotAction; message: string };
      }>(
        `/autopilot/actions/${actionId}/confirm`,
        overrideReason ? { override_reason: overrideReason } : undefined
      );
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-summary', tenantId] });
    },
  });
}

/**
 * Approve multiple actions at once.
 */
export function useApproveAllActions(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (actionIds?: string[]) => {
      const response = await apiClient.post<{ data: { approved_count: number } }>(
        `/autopilot/actions/approve-all`,
        actionIds ? { action_ids: actionIds } : undefined
      );
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-summary', tenantId] });
    },
  });
}

/**
 * Dismiss an action.
 */
export function useDismissAction(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (actionId: string) => {
      const response = await apiClient.post<{ data: { action: AutopilotAction } }>(
        `/autopilot/actions/${actionId}/dismiss`
      );
      return response.data.data.action;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
    },
  });
}

// =============================================================================
// Enforcement settings + emergency stop (freeze)
// =============================================================================

export interface EnforcementSettings {
  /** Guardrails toggle. False DISABLES enforcement checks (loosens control). */
  enforcement_enabled: boolean;
  /** Emergency stop. True halts ALL autopilot execution regardless of mode. */
  autopilot_frozen: boolean;
  default_mode: 'advisory' | 'soft_block' | 'hard_block';
}

/**
 * Read the tenant's enforcement settings, including the emergency-stop
 * (`autopilot_frozen`) state that powers the kill switch UI.
 */
export function useEnforcementSettings(tenantId: number) {
  return useQuery({
    queryKey: ['enforcement-settings', tenantId],
    queryFn: async () => {
      const response = await apiClient.get<{
        data: { settings: EnforcementSettings };
      }>(`/autopilot/enforcement/settings`);
      return response.data.data.settings;
    },
    enabled: !!tenantId,
    staleTime: 30 * 1000,
    refetchInterval: 30 * 1000,
  });
}

/**
 * Toggle the autopilot emergency stop. Freezing (`frozen: true`) halts all
 * execution — the apply worker skips approved actions and the
 * approve/confirm endpoints refuse — so nothing reaches a platform until
 * autopilot is resumed. Invalidates settings + action/status queries so the
 * whole autopilot surface reflects the new state immediately.
 */
export function useSetFreeze(tenantId: number) {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async ({ frozen, reason }: { frozen: boolean; reason?: string }) => {
      const response = await apiClient.post<{
        data: { autopilot_frozen: boolean; message: string };
      }>(`/autopilot/enforcement/freeze`, { frozen, reason });
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['enforcement-settings', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-status', tenantId] });
      queryClient.invalidateQueries({ queryKey: ['autopilot-actions', tenantId] });
    },
  });
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Get human-readable action type label.
 */
export function getActionTypeLabel(type: ActionType): string {
  const labels: Record<ActionType, string> = {
    budget_increase: 'Budget Increase',
    budget_decrease: 'Budget Decrease',
    pause_campaign: 'Pause Campaign',
    pause_adset: 'Pause Ad Set',
    pause_creative: 'Pause Creative',
    enable_campaign: 'Enable Campaign',
    enable_adset: 'Enable Ad Set',
    enable_creative: 'Enable Creative',
    bid_increase: 'Bid Increase',
    bid_decrease: 'Bid Decrease',
  };
  return labels[type] || type;
}

/**
 * Get status color for UI.
 */
export function getActionStatusColor(status: ActionStatus): string {
  const colors: Record<ActionStatus, string> = {
    queued: 'yellow',
    pending_approval: 'orange',
    approved: 'blue',
    applied: 'green',
    failed: 'red',
    dismissed: 'gray',
  };
  return colors[status] || 'gray';
}

/**
 * Get status label.
 */
export function getActionStatusLabel(status: ActionStatus): string {
  const labels: Record<ActionStatus, string> = {
    queued: 'Pending Approval',
    pending_approval: 'Needs Confirmation',
    approved: 'Approved',
    applied: 'Applied',
    failed: 'Failed',
    dismissed: 'Dismissed',
  };
  return labels[status] || status;
}

/**
 * Get platform icon.
 */
export function getPlatformIcon(platform: string): string {
  const icons: Record<string, string> = {
    meta: '📘',
    google: '🔴',
    tiktok: '🎵',
    snapchat: '👻',
  };
  return icons[platform] || '📊';
}

File: backend/app/api/v1/endpoints/autopilot.py
# =============================================================================
# ADs Growth System - Autopilot API Router
# =============================================================================
"""
API endpoints for Autopilot features:
- Action queue management
- Approval workflow
- Action execution
"""

import logging
from datetime import date, datetime, timezone
from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Body, Depends, HTTPException, Query, Request
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.deps import get_current_user
from app.autopilot.service import ActionStatus, ActionType, AutopilotService
from app.db.session import get_async_session
from app.features.service import can_access_feature, get_org_features
from app.schemas.response import APIResponse

logger = logging.getLogger(__name__)

# SECURITY (STRAT-SC-001/C3): the old per-org request guards were the only
# auth on these routes; deleted in the de-tenanting sweep, so real auth is
# enforced router-wide here.
router = APIRouter(
    prefix="/autopilot",
    tags=["autopilot"],
    dependencies=[Depends(get_current_user)],
)


def _dispatch_single_action(action_id: str, user_id: Optional[int]) -> None:
    """Dispatch execution of a single approved action to the Celery worker.

    Import is local to avoid a circular import at module load (the task module
    imports app.autopilot.service). Broker failures are swallowed so a transient
    dispatch problem never rolls back an already-committed approval — the
    every-5-min ``apply_actions_queue`` sweep is the backstop that catches it.
    """
    try:
        from app.tasks.apply_actions_queue import apply_single_action

        apply_single_action.delay(action_id, user_id)
    except Exception:  # pragma: no cover - defensive, backstop sweep recovers
        logger.exception(
            "Failed to dispatch apply_single_action for action %s; "
            "the scheduled sweep will retry",
            action_id,
        )


def _dispatch_rollback(action_id: str, user_id: Optional[int]) -> None:
    """Dispatch a rollback of an applied action to the Celery worker (TRUST-006).

    Local import to avoid a circular import at module load; broker failures are
    swallowed so a transient dispatch problem doesn't 500 an already-validated
    rollback request.
    """
    try:
        from app.tasks.apply_actions_queue import rollback_action

        rollback_action.delay(action_id, user_id)
    except Exception:  # pragma: no cover - defensive
        logger.exception("Failed to dispatch rollback_action for action %s", action_id)


def _dispatch_apply_queue() -> None:
    """Dispatch execution of all approved actions (used by approve-all).
    Same defensive contract as ``_dispatch_single_action``."""
    try:
        from app.tasks.apply_actions_queue import apply_actions_queue

        apply_actions_queue.delay()
    except Exception:  # pragma: no cover - defensive, backstop sweep recovers
        logger.exception(
            "Failed to dispatch apply_actions_queue; the scheduled sweep will retry"
        )


async def _require_not_frozen(db: AsyncSession) -> None:
    """Raise 409 when autopilot is frozen (emergency stop).

    Guards the execution-triggering endpoints (approve / confirm /
    approve-all) so an operator gets an explicit error instead of an action
    that is silently deferred. The apply worker enforces the same freeze
    authoritatively; this is the UX-layer echo so nothing looks "approved
    but stuck".
    """
    from app.autopilot.enforcer import AutopilotEnforcer

    if await AutopilotEnforcer(db).is_frozen():
        raise HTTPException(
            status_code=409,
            detail=(
                "Autopilot is frozen (emergency stop). "
                "Resume autopilot before executing actions."
            ),
        )


# =============================================================================
# Request/Response Models
# =============================================================================


class QueueActionRequest(BaseModel):
    """Request to queue a new action."""

    action_type: str = Field(..., description="Type of action")
    entity_type: str = Field(
        ..., description="Type of entity (campaign, adset, creative)"
    )
    entity_id: str = Field(..., description="Platform entity ID")
    entity_name: str = Field(..., description="Human-readable name")
    platform: str = Field(
        ..., description="Ad platform (meta, google, tiktok, snapchat)"
    )
    action_json: Dict[str, Any] = Field(..., description="Full action details")
    before_value: Optional[Dict[str, Any]] = Field(
        None, description="Current value before change"
    )


class ApproveActionsRequest(BaseModel):
    """Request to approve multiple actions."""

    action_ids: List[str] = Field(..., description="List of action UUIDs to approve")


class ApproverInfo(BaseModel):
    """User who accepted an action, for accountability display."""

    id: int
    name: Optional[str] = None
    department: Optional[str] = None


class ActionResponse(BaseModel):
    """Serialized action for API response."""

    id: str
    date: str
    action_type: str
    entity_type: str
    entity_id: str
    entity_name: Optional[str]
    platform: str
    action_json: Dict[str, Any]
    before_value: Optional[Dict[str, Any]]
    after_value: Optional[Dict[str, Any]]
    status: str
    created_at: str
    approved_at: Optional[str]
    applied_at: Optional[str]
    error: Optional[str]
    requires_confirmation: bool = False
    confirmation_token: Optional[str] = None
    approved_by: Optional[ApproverInfo] = None


# =============================================================================
# Helper Functions
# =============================================================================


def _approver_info(action) -> Optional[ApproverInfo]:
    """Serialize the eager-loaded approver, if any.

    Only touches the relationship when the query eager-loaded it
    (selectinload in AutopilotService); a lazy load here would raise
    MissingGreenlet under the async session. full_name is PII-encrypted
    at rest, so decrypt for display.
    """
    from sqlalchemy import inspect as sa_inspect

    from app.core.security import decrypt_pii

    if action.approved_by_user_id is None:
        return None
    if "approved_by" in sa_inspect(action).unloaded:
        return None
    approver = action.approved_by
    if approver is None:  # user row deleted (FK is SET NULL on hard delete)
        return None

    name: Optional[str] = None
    if approver.full_name:
        try:
            name = decrypt_pii(approver.full_name)
        except (ValueError, TypeError):
            name = None

    return ApproverInfo(
        id=approver.id,
        name=name,
        department=approver.department,
    )


def action_to_response(action) -> ActionResponse:
    """Convert database action to API response."""
    import json

    return ActionResponse(
        id=str(action.id),
        date=action.date.isoformat(),
        action_type=action.action_type,
        entity_type=action.entity_type,
        entity_id=action.entity_id,
        entity_name=action.entity_name,
        platform=action.platform,
        action_json=json.loads(action.action_json) if action.action_json else {},
        before_value=json.loads(action.before_value) if action.before_value else None,
        after_value=json.loads(action.after_value) if action.after_value else None,
        status=action.status,
        created_at=action.created_at.isoformat() if action.created_at else None,
        approved_at=action.approved_at.isoformat() if action.approved_at else None,
        applied_at=action.applied_at.isoformat() if action.applied_at else None,
        error=action.error,
        requires_confirmation=bool(
            action.status == ActionStatus.PENDING_APPROVAL.value
            and getattr(action, "confirmation_token", None)
        ),
        confirmation_token=getattr(action, "confirmation_token", None),
        approved_by=_approver_info(action),
    )


# =============================================================================
# Endpoints
# =============================================================================


@router.get("/status", response_model=APIResponse[Dict[str, Any]])
async def get_autopilot_status(
    request: Request,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get current autopilot status and configuration.

    Returns:
    - autopilot_level: Current level (0=suggest, 1=guarded, 2=approval)
    - pending_actions: Number of actions awaiting approval
    - caps: Budget/action caps for guarded mode
    - enabled: Whether autopilot is enabled
    """
    features = await get_org_features(db)

    service = AutopilotService(db)
    summary = await service.get_action_summary(days=1)

    from app.features.flags import get_autopilot_caps

    caps = get_autopilot_caps()

    return APIResponse(
        success=True,
        data={
            "autopilot_level": features.get("autopilot_level", 0),
            "autopilot_level_name": {
                0: "Suggest Only",
                1: "Guarded Auto",
                2: "Approval Required",
            }.get(features.get("autopilot_level", 0), "Unknown"),
            "pending_actions": summary["pending_approval"],
            "caps": caps,
            "enabled": features.get("autopilot_level", 0) > 0,
        },
    )


@router.get("/actions", response_model=APIResponse[Dict[str, Any]])
async def get_actions(
    request: Request,
    target_date: Optional[date] = Query(default=None, alias="date"),
    status: Optional[str] = Query(default=None),
    platform: Optional[str] = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Get actions.

    Query params:
    - date: Filter by date
    - status: Filter by status (queued, approved, applied, failed, dismissed)
    - platform: Filter by platform
    - limit: Max results (default 50)
    """
    service = AutopilotService(db)
    actions = await service.get_queued_actions(
        target_date=target_date,
        status=status,
        platform=platform,
        limit=limit,
    )

    return APIResponse(
        success=True,
        data={
            "actions": [action_to_response(a).dict() for a in actions],
            "count": len(actions),
            "filters": {
                "date": target_date.isoformat() if target_date else None,
                "status": status,
                "platform": platform,
            },
        },
    )


@router.get("/actions/summary", response_model=APIResponse[Dict[str, Any]])
async def get_actions_summary(
    request: Request,
    days: int = Query(default=7, ge=1, le=30),
    db: AsyncSession = Depends(get_async_session),
):
    """Get summary of actions over the past N days."""
    service = AutopilotService(db)
    summary = await service.get_action_summary(days=days)

    return APIResponse(success=True, data=summary)


@router.get("/outcomes/summary", response_model=APIResponse[Dict[str, Any]])
async def get_outcomes_summary(
    request: Request,
    period: str = Query(default="7d", pattern="^(24h|7d|30d)$"),
    db: AsyncSession = Depends(get_async_session),
):
    """
    Aggregate the dollar value Stratum's autopilot has delivered
    over a period. Powers the outcome-triggered upgrade nudge
    on the dashboard home (see frontend/src/components/billing/
    OutcomeNudge.tsx).

    Phase A: returns 0/0/0 for everyone because the estimator is a stub.
    Phase B: real counterfactual numbers with conservative_factor=0.5.
    """
    from app.services.autopilot import get_outcome_summary

    summary = await get_outcome_summary(db, period=period)
    return APIResponse(success=True, data=summary.to_dict())


@router.post("/actions", response_model=APIResponse[Dict[str, Any]])
async def queue_action(
    request: Request,
    body: QueueActionRequest,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Queue a new action.
    Actions will be auto-executed or require approval based on autopilot level.
    """
    # Get current user ID from request state if available
    user_id = getattr(request.state, "user_id", None)

    service = AutopilotService(db)

    # Get autopilot level
    features = await get_org_features(db)
    autopilot_level = features.get("autopilot_level", 0)

    # Check if action can be auto-executed
    can_auto, reason = service.can_auto_execute(
        autopilot_level,
        body.action_type,
        body.action_json,
    )

    action = await service.queue_action(
        action_type=body.action_type,
        entity_type=body.entity_type,
        entity_id=body.entity_id,
        entity_name=body.entity_name,
        platform=body.platform,
        action_json=body.action_json,
        before_value=body.before_value,
        created_by_user_id=user_id,
    )

    # Auto-approve if allowed
    if can_auto:
        action = await service.approve_action(action.id, user_id or 0)
        # Auto-approved actions execute immediately (trust gate already passed).
        if action:
            _dispatch_single_action(str(action.id), user_id)

    return APIResponse(
        success=True,
        data={
            "action": action_to_response(action).dict(),
            "auto_approved": can_auto,
            "requires_approval": not can_auto,
            "reason": reason,
        },
    )


@router.post("/actions/{action_id}/approve", response_model=APIResponse[Dict[str, Any]])
async def approve_action(
    request: Request,
    action_id: str,
    db: AsyncSession = Depends(get_async_session),
):
    """Approve a queued action for execution."""
    user_id = getattr(request.state, "user_id", None)
    if not user_id:
        raise HTTPException(status_code=401, detail="User authentication required")

    try:
        uuid_id = UUID(action_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid action ID format")

    # Emergency stop — refuse to approve-for-execution while frozen.
    await _require_not_frozen(db)

    service = AutopilotService(db)
    action = await service.approve_action(uuid_id, user_id)

    if not action:
        raise HTTPException(
            status_code=404, detail="Action not found or already processed"
        )

    # Dispatch execution now that the action is approved and committed.
    _dispatch_single_action(str(action.id), user_id)

    return APIResponse(
        success=True,
        data={
            "action": action_to_response(action).dict(),
            "message": "Action approved for execution",
        },
    )


@router.post(
    "/actions/{action_id}/rollback", response_model=APIResponse[Dict[str, Any]]
)
async def rollback_applied_action(
    request: Request,
    action_id: str,
    db: AsyncSession = Depends(get_async_session),
):
    """Roll back a previously applied autopilot action (TRUST-006).

    Dispatches the inverse action (restoring the entity's before_value). Only an
    applied, reversible action can be rolled back.
    """
    user_id = getattr(request.state, "user_id", None)
    if not user_id:
        raise HTTPException(status_code=401, detail="User authentication required")

    try:
        uuid_id = UUID(action_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid action ID format")

    await _require_not_frozen(db)

    service = AutopilotService(db)
    action = await service.get_action_by_id(uuid_id)
    if not action:
        raise HTTPException(status_code=404, detail="Action not found")
    if action.status != ActionStatus.APPLIED.value:
        raise HTTPException(
            status_code=409,
            detail=f"Only applied actions can be rolled back (status={action.status})",
        )

    _dispatch_rollback(str(action.id), user_id)
    return APIResponse(
        success=True,
        data={"action_id": str(action.id), "message": "Rollback initiated"},
    )


class ConfirmActionRequest(BaseModel):
    """Request to confirm a soft-blocked action."""

    override_reason: Optional[str] = Field(
        None, description="Why the operator is overriding the soft-block"
    )


@router.post("/actions/{action_id}/confirm", response_model=APIResponse[Dict[str, Any]])
async def confirm_soft_blocked_action(
    request: Request,
    action_id: str,
    body: Optional[ConfirmActionRequest] = None,
    db: AsyncSession = Depends(get_async_session),
):
    """Confirm a soft-blocked action and execute it.

    When the execution-path enforcement gate soft-blocks an approved
    action, it returns the action to PENDING_APPROVAL with a one-time
    confirmation token. This endpoint consumes that token (audit-logging
    the override), stamps the action as operator-confirmed, re-approves
    it, and dispatches execution. The gate lets the confirmed action
    through; hard-blocks remain absolute.

    If the stored token has expired, the action is re-dispatched so the
    gate re-evaluates and mints a fresh token.
    """
    user_id = getattr(request.state, "user_id", None)
    if not user_id:
        raise HTTPException(status_code=401, detail="User authentication required")

    try:
        uuid_id = UUID(action_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid action ID format")

    # Emergency stop — frozen autopilot cannot override soft-blocks either.
    await _require_not_frozen(db)

    service = AutopilotService(db)
    action = await service.get_action_by_id(uuid_id)
    if not action:
        raise HTTPException(status_code=404, detail="Action not found")

    if action.status != ActionStatus.PENDING_APPROVAL.value:
        raise HTTPException(
            status_code=409,
            detail=f"Action status is {action.status}; only soft-blocked "
            "(pending_approval) actions can be confirmed",
        )

    if not action.confirmation_token:
        raise HTTPException(
            status_code=409,
            detail="Action has no pending confirmation token",
        )

    from app.autopilot.enforcer import AutopilotEnforcer

    enforcer = AutopilotEnforcer(db)
    success, error = await enforcer.confirm_action(
        confirmation_token=action.confirmation_token,
        user_id=user_id,
        override_reason=body.override_reason if body else None,
    )

    now = datetime.now(timezone.utc)

    if not success:
        # Expired/invalid token: clear it and re-dispatch so the gate
        # re-evaluates and mints a fresh token for the operator.
        action.confirmation_token = None
        action.status = ActionStatus.APPROVED.value
        action.error = None
        await db.commit()
        _dispatch_single_action(str(action.id), user_id)
        return APIResponse(
            success=False,
            data={
                "action": action_to_response(action).dict(),
                "message": (
                    f"Confirmation failed ({error}). The action was "
                    "re-submitted for enforcement evaluation; a fresh "
                    "confirmation token will be issued if it is "
                    "soft-blocked again."
                ),
            },
        )

    action.enforcement_confirmed_at = now
    action.enforcement_confirmed_by_user_id = user_id
    action.confirmation_token = None
    action.status = ActionStatus.APPROVED.value
    action.approved_by_user_id = action.approved_by_user_id or user_id
    action.approved_at = action.approved_at or now
    action.error = None
    await db.commit()
    await db.refresh(action)

    # Dispatch execution now that the override is committed.
    _dispatch_single_action(str(action.id), user_id)

    return APIResponse(
        success=True,
        data={
            "action": action_to_response(action).dict(),
            "message": "Override confirmed - action dispatched for execution",
        },
    )


@router.post("/actions/approve-all", response_model=APIResponse[Dict[str, Any]])
async def approve_all_actions(
    request: Request,
    body: Optional[ApproveActionsRequest] = None,
    db: AsyncSession = Depends(get_async_session),
):
    """Approve multiple queued actions at once."""
    user_id = getattr(request.state, "user_id", None)
    if not user_id:
        raise HTTPException(status_code=401, detail="User authentication required")

    # Emergency stop — refuse bulk approve-for-execution while frozen.
    await _require_not_frozen(db)

    service = AutopilotService(db)

    action_ids = None
    if body and body.action_ids:
        try:
            action_ids = [UUID(aid) for aid in body.action_ids]
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid action ID format")

    count = await service.approve_all_queued(user_id, action_ids)

    # Dispatch the apply sweep so the just-approved actions execute.
    if count:
        _dispatch_apply_queue()

    return APIResponse(
        success=True,
        data={
            "approved_count": count,
            "message": f"Approved {count} actions for execution",
        },
    )


@router.post("/actions/{action_id}/dismiss", response_model=APIResponse[Dict[str, Any]])
async def dismiss_action(
    request: Request,
    action_id: str,
    db: AsyncSession = Depends(get_async_session),
):
    """Dismiss a queued action (won't be executed)."""
    user_id = getattr(request.state, "user_id", None)
    if not user_id:
        raise HTTPException(status_code=401, detail="User authentication required")

    try:
        uuid_id = UUID(action_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid action ID format")

    service = AutopilotService(db)
    action = await service.dismiss_action(uuid_id, user_id)

    if not action:
        raise HTTPException(
            status_code=404, detail="Action not found or already processed"
        )

    return APIResponse(
        success=True,
        data={
            "action": action_to_response(action).dict(),
            "message": "Action dismissed",
        },
    )


@router.get("/actions/{action_id}", response_model=APIResponse[Dict[str, Any]])
async def get_action(
    request: Request,
    action_id: str,
    db: AsyncSession = Depends(get_async_session),
):
    """Get details of a specific action."""
    try:
        uuid_id = UUID(action_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid action ID format")

    service = AutopilotService(db)
    action = await service.get_action_by_id(uuid_id)

    if not action:
        raise HTTPException(status_code=404, detail="Action not found")

    return APIResponse(
        success=True,
        data={"action": action_to_response(action).dict()},
    )
